design_docs/protostar-architecture.md

```diff
@@ -0,0 +1,63 @@
+# Protostar modular architecture
+
+This document proposes the idea how to modularize Protostar. Goals of the modularization are
+- Allow to introduce better ownerships of the codebase parts
+- Make workflow asynchronous between smaller sub-teams operating on contracts between modules
+- Allow incremental rewrite to Rust
+- Remove compilation responsibility from Protostar codebase
+
+
+
+## General overview
+First decomposition of Protostar will introduce 3 modules
+
+- Protostar Gateway
+- Protostar Starknet
+- Protostar Testing backend
+
+Each of the modules would be a separate poetry package (open question?), to simplify scenarios of updating different dependencies. Example of such scenario, Protostar-starknet relies on starknet.py and devnet but testing backend does not. As a final goal everything would be under one cargo workspace.
+![Diagram](img/diagram.png)
+
+## Modules
+### Protostar Gateway
+This module will be a communication layer of Protostar with outer world, namely Scarb. Its responsibility will be to provide Scarb with a correct cairo compiler (or dependencies in the future) and to fetch configuration from Scarb config. This module will be routing commands as requests to other services. Output formatting will be responsibility of the Gateway
+
+
+### Testing backend
+This module will be responsible for resolving a transaction
+```
+Input: List of test suites and their code, testing backend config
+Output: Stream of test results + summary as a final element
+```
+
+Anything outside of this transaction will be encapsulated in the backend.
+
+
+### Starknet CLI
+This is a facade over starknet.py SDK, we can migrate it to Rust very soon and use Starknet.rs instead if its capabilities are sufficient.
+
+
+## Communication
+Communication layer between modules will be JSON, simillar to the current Protostar `--json` to make it language-independent. We can do communication either through FFI or calling as sub-processes, this is an open question.
```

Comment:
I know that FFI is causing a lot of pain right now and might be harder to work with, WDYT?

Reply:
Do you really need two-way live communication between gateway and deeper modules? I thought it would be more like command execution? Why not simply pass params via arguments/env/argfile, pipe stdin/stdout/stderr and wait for execution?

Reply:
that's a good point, we have to probably think about the implications though

Reply:
@mkaput I know that was a first idea, on the other hand to way communication removes formatting responsibility form specific teams which makes it much easier to keep consistent. I still consider it as an open question

Reply:
The problem is that human-readable and JSON formats are not in sub/superset relation. You would basically have to invent a third format for transport. I don't think it's worth it over copy-paste of a rather simple mechanism that is present currently in Protostar.

Reply:
I wouldn't overcomplicate it - I think the solution proposed by @mkaput is just fine. End users doesn't really care how it is done and for us (developers) the cons outweigh the pros imo, at least for now.